### PR TITLE
Script adding a GCR staging repo

### PR DESCRIPTION
**Pull request by thockin**

### CODE REVIEW

**Comment by thockin** [comment]
As promised on Friday, the script.

**Comment by dims** [comment]
Related to https://github.com/kubernetes/k8s.io/issues/158

File: gcr/add-staging-repo.sh [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+color 6 "Configuring billing for ${PROJECT}"
+gcloud beta billing projects link "${PROJECT}" \
+    --billing-account "${BILLING}"
+
+# Grant project viewer so the UI will work.
+color 6 "Granting project viewer to ${ADMINS}"
+gcloud \
+    projects add-iam-policy-binding "${PROJECT}" \
+    --member "group:${ADMINS}" \
+    --role roles/viewer
```
Comment (javier-b-perez) [inline gcr/add-staging-repo.sh]: do we just want admins as 'viewers'? what about 'owners'?

File: gcr/add-staging-repo.sh [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+gsutil iam ch "group:${ADMINS}:legacyBucketOwner" "gs://${BUCKET}"
+
+# Make the bucket publicly readable.
+color 6 "Making the bucket public"
+gsutil iam ch allUsers:objectViewer "gs://${BUCKET}"
+
+# Grant repo writers access to write.
+color 6 "Granting bucket objectAdmin to ${WRITERS}"
+gsutil iam ch "group:${WRITERS}:objectAdmin" "gs://${BUCKET}"
+color 6 "Granting bucket legacyBucketReader to ${WRITERS}"
```
Comment (javier-b-perez) [inline gcr/add-staging-repo.sh]: do we need 'legacyBucketReader'?
are we going to store GCS buckets as well or just GCR?

File: gcr/add-staging-repo.sh (hunk 1 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: They only need project viewer for the UI to work.  They have bucket admin for the specific buckets.

We're trying to keep untra-limited perms when possible.

File: gcr/add-staging-repo.sh (hunk 2 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: without buckets.list it didn't work, and there doesn't seem to be a role that includes that permission that isn't "legacy" (and I didn't want to make a custom role, though we could, I guess...)

File: gcr/add-staging-repo.sh (hunk 1 above)
Comment (javier-b-perez) [reply gcr/add-staging-repo.sh]: who is going to manage the projects? just who create them?

File: gcr/add-staging-repo.sh (hunk 2 above)
Comment (javier-b-perez) [reply gcr/add-staging-repo.sh]: you are right, the other alternative is 'roles/storage.admin'

**Comment by javier-b-perez** [comment]
LGTM

as side note, we should also document somewhere who is going to own the groups like 'k8s-infra-gcr-staging-${REPO}@googlegroups.com', probably another google group?

File: gcr/add-staging-repo.sh (hunk 1 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: the set of people who can create projects is very small and should, itself be governed by a googlegroup

File: gcr/add-staging-repo.sh (hunk 2 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: which is too broad

**Comment by dims** [comment]
/assign @justinsb 

Justin, can you please peek?

**Comment by thockin** [comment]
I am working on a quick followup commit to age-out old data

**Comment by thockin** [comment]
PTAL at last commit, too

File: gcr/add-staging-repo.sh [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+# The group that can write to this staging repo.
+WRITERS="k8s-infra-gcr-staging-${REPO}@googlegroups.com"
+
+# The GCP org stuff needed to turn it all on.
+ORG="758905017065" # kubernetes.io
+BILLING="018801-93540E-22A20E"
+
+# Make the project, if needed
+if ! gcloud projects describe "${PROJECT}" >/dev/null 2>&1; then
```
Comment (justinsb) [inline gcr/add-staging-repo.sh]: What happens if this project name happens to be an unrelated 3rd party?

File: gcr/add-staging-repo.sh [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+gcloud --project "${PROJECT}" \
+    services enable containerregistry.googleapis.com
+
+# Push an image to trigger the bucket to be created
+color 6 "Activating the registry bucket"
+docker pull k8s.gcr.io/pause
+docker tag k8s.gcr.io/pause "gcr.io/${PROJECT}/pause"
+docker push "gcr.io/${PROJECT}/pause"
+gcloud --project "${PROJECT}" \
+    container images delete --quiet "gcr.io/${PROJECT}/pause:latest"
```
Comment (justinsb) [inline gcr/add-staging-repo.sh]: Scary!  Worth tagging `gcr.io/${PROJECT}/ceci-nest-pas-une-image` or some other thing that isn't important?

**Comment by justinsb** [comment]
Two paranoid suggestions, but lgtm

/lgtm

File: gcr/add-staging-repo.sh (hunk 3 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: You won't be able to list those.  I don't see a way to only list projects under a specific org - do you know of one?  Assuming you have a project with this name pattern in your scope, something else will fail.  Or if you have project admin, maybe it won't fail.

We could require this create to happen, which makes the scripts not re-runnable.  We could add flags to say --project-exists or something.  Is that worth the effort?

File: gcr/add-staging-repo.sh (hunk 4 above)
Comment (thockin) [reply gcr/add-staging-repo.sh]: hahah, sure.  In next push

**Comment by thockin** [comment]
New push is up with image name change.  Other is unresolved. Launch and iterate?

**Comment by dims** [comment]
thanks for adding the age-out. +1 to launch/iterate

/approve

leaving `/lgtm` for @justinsb

**Comment by thockin** [comment]
We could check the project existence with:

gcloud projects describe k8s-staging-coredns --flatten='parent[]'
--format='csv[no-heading](type, id)' | grep ^organization, | cut -f2 -d,
and match that against the

**Comment by thockin** [comment]
Pushed with that added.  @justinsb should be placated now...

**Comment by thockin** [comment]
ping for lgtm

**Pull request merged**